File: search/binary_search/go/binary_search.go
package binarysearch

// BinarySearch performs a binary search on a given sorted integer slice
// 'sortedSlice' for an integer value 'value'. If this value can be found
// in the sorted slice, the index of the first occurence will be returned,
// together with a boolean 'true' to indicate, that the search was successful.
// If the given value is not contained in the slice, an arbitrary index can
// be returned, together with a boolean 'false' to indicate, that the
// given value is not contained in the slice.
func BinarySearch(value int, sortedSlice []int) (index int, found bool) {
	low, high := 0, len(sortedSlice)-1
	for low <= high {
		mid := (low + high) / 2
		if sortedSlice[mid] < value {
			low = mid + 1
		} else if sortedSlice[mid] > value {
			high = mid - 1
		} else if low != mid {
			high = mid
		} else {
			return mid, true
		}
	}
	return -1, false
}

File: search/binary_search/Rust/binary_search.rs
fn binary_search(arr :Vec<i32>, n :i32) -> i32{
    let mut low = 0;
    let mut high = arr.len()-1;
   loop {
        let mid = (low + high) / 2;
        if n == arr[mid] {
            return mid as i32;
        }else if n > arr[mid] {
            low = mid + 1;
        }else {
            high = mid -1;
        }
        if low >= high {
            break;
        }
    }
    -1
}

fn main() {
    let arr = vec![0, 4, 5, 8, 12, 14, 20, 34, 55, 60, 84, 111, 134, 154, 188, 189, 200];
    let num = 8;
    let index = binary_search(arr, num);
    if index != -1 {
        println!("{} found at index {}",num,index);
    }else {
        println!("{} not found!", num);
    }
}